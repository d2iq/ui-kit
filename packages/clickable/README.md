# Clickable

The Clickable is a higher order presentational component, children get the necessary a11y attributes added.

Use this whenever you want to bind an action on click onto a HTML element. The component will add the appropiate role attribute, add a tabIndex, add a focus state and bind the action to `onClick` and `onKeyPress`. The `onKeyPress` has a wrapping function which filters out eveerything except the space key and enter key events. This prevents that the function gets triggered by accident.
