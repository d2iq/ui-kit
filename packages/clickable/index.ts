export { default as Clickable } from "./clickable";
