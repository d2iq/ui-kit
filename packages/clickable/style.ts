import { css } from "@emotion/css";

export const outline = css`
  &:focus {
    outline: 0;
  }
`;

export const pointer = css`
  cursor: pointer;
`;
