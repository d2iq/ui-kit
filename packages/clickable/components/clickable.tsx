import { cx } from "@emotion/css";
import * as React from "react";
import { outline, pointer } from "../style";

export interface ClickableProps {
  /**
   * Children should be a HTML element.
   */
  children: React.ReactElement<HTMLElement>;
  /**
   * Action is a event handler for the onClick and onKeypress events
   */
  action: (event?: React.SyntheticEvent<HTMLElement>) => void;
  /**
   * The tabIndex is passed down and is the same as the native tabIndex
   */
  tabIndex?: number | string;
  /**
   * ARIA role of the clickable element
   */
  role?: string;
  /**
   * Whether or not to reset the :focus outline style
   */
  disableFocusOutline?: boolean;
  /**
   * human-readable selector used for writing tests
   */
  ["data-cy"]?: string;
}

export class Clickable extends React.PureComponent<ClickableProps, {}> {
  public static defaultProps: Partial<ClickableProps> = {
    tabIndex: -1,
    role: "button",
    disableFocusOutline: false
  };

  constructor(props: ClickableProps) {
    super(props);
    this.handleKeyPress = this.handleKeyPress.bind(this);
  }

  public render() {
    const {
      children,
      action,
      tabIndex,
      role,
      disableFocusOutline,
      "data-cy": dataCy
    } = this.props;
    const { className = "" } = children.props;

    return React.cloneElement(React.Children.only(children), {
      onClick: action,
      className: cx(className, pointer, { [outline]: disableFocusOutline }),
      role,
      tabIndex,
      onKeyPress: this.handleKeyPress,
      ["data-cy"]: dataCy
    });
  }

  public handleKeyPress(event: React.KeyboardEvent<HTMLElement>): void {
    if (event.key === " " || event.key === "Enter") {
      this.props.action(event);
    }
  }
}

export default Clickable;
