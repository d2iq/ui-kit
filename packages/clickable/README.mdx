---
name: Clickable
menu: Components
route: /clickable
---

import { Playground, PropsTable } from "docz";
import { default as Clickable } from "./components/clickable";

# Clickable

- [Clickable](#clickable)
  - [Clickable Component](#clickable-component)
  - [Props](#props)
  - [Example](#example)

## Clickable Component

The Clickable is a higher order presentational component, children get the necessary a11y attributes added.

Use this whenever you want to bind an action on click onto a HTML element. The component will add the appropiate role attribute, add a tabIndex, add a focus state and bind the action to `onClick` and `onKeyPress`. The `onKeyPress` has a wrapping function which filters out eveerything except the space key and enter key events. This prevents that the function gets triggered by accident.

## Props

<PropsTable of={Clickable} />

## Example

<Playground>
  <Clickable action={() => alert("action trigger")} tabIndex="0">
    <span>Click me!</span>
  </Clickable>
</Playground>
