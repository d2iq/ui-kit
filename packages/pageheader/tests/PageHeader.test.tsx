import React from "react";
import { render } from "@testing-library/react";

import { PageHeader } from "../";
import { Clickable } from "../../clickable";

describe("PageHeader", () => {
  it("default", () => {
    const action = () => 1;
    const { asFragment } = render(
      <PageHeader
        breadcrumbElements={[
          <span key="One">One</span>,
          <span key="Two">Two</span>
        ]}
        actions={[
          <Clickable key="Action1" action={action}>
            <div>Action 1</div>
          </Clickable>
        ]}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
