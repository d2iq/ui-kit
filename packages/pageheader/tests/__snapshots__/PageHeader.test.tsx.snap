// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader default 1`] = `
<div
  class="css-fon3qn"
>
  <div
    class="css-f1o908"
    data-cy="pageHeader"
  >
    <div
      class="css-qi8t7c"
    >
      <nav
        class="css-cldwuj"
      >
        <div
          class="css-zsjup2"
          data-cy="breadcrumb"
        >
          <span>
            One
          </span>
        </div>
        <div
          class="css-10fubqe"
          data-cy="breadcrumb"
        >
          <svg
            aria-label="system-caret-right icon"
            class="css-jbdp75"
            data-cy="icon"
            height="16"
            preserveAspectRatio="xMinYMin meet"
            role="img"
            viewBox="0 0 16 16"
            width="16"
          >
            <use
              href="#system-caret-right"
            />
          </svg>
        </div>
        <div
          class="css-10fubqe"
          data-cy="breadcrumb"
        >
          <span>
            Two
          </span>
        </div>
      </nav>
    </div>
    <div
      class="css-48ylac"
    >
      <ul
        class="css-eozmc"
        data-cy="pageHeader-actions"
      >
        <li
          class="css-2h2dvf"
        >
          <div
            class="css-1h5x3dy"
            role="button"
            tabindex="-1"
          >
            Action 1
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
